api/envoy/extensions/http/early_header_mutation/regex_mutation/v3/regex_mutation.proto

```diff
@@ -0,0 +1,25 @@
+syntax = "proto3";
+
+package envoy.extensions.http.early_header_mutation.regex_mutation.v3;
+
+import "envoy/config/common/mutation_rules/v3/mutation_rules.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.http.early_header_mutation.regex_mutation.v3";
+option java_outer_classname = "RegexMutationProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/http/early_header_mutation/regex_mutation/v3;regex_mutationv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Custom early header mutation extension]
+// [#extension: envoy.http.early_header_mutation.regex_mutation]
+
+// This extension allows for custom early header mutation by the regex match and rewrite.
+// It is configured with a list of
+// :ref:`header mutation <envoy_v3_api_msg_config.common.mutation_rules.v3.HeaderMutation>`.
+message RegexMutation {
+  repeated config.common.mutation_rules.v3.HeaderMutation header_mutations = 1
+      [(validate.rules).repeated = {min_items: 1}];
+}
```

Comment:
Honestly I'm on the fence as to whether this should be an extension or not. It seems like there could just be a list of early header mutations that live within the HCM config? @envoyproxy/api-shepherds @alyssawilk wdyt? It seems like a fair amount of complexity and it's not clear to me that it really needs full customization for any known use case.

If we do keep the extension, I would not call this the RegexMutation. I would call it something more generic that can just operate on the HeaderMutation structure which could eventually do other things per my oneof comment.

Reply:
I still don't grok why we can't do this as a filter and just clear the route cache, but if we're going to add it I'd prefer we add it as an extension as I can only imagine it growing over time

Reply:
I think the issue is that a) there may be significant perf cost for reloading the route and b) some other stuff like tracing pivots on this before filters run.

An extension is OK with me but let's try to make the default version more generic.

Reply:
yeah agree that's an interesting option to explore. I'm not sure how hard that would be to implement @wbpcode? In general I agree less extension points is better if we don't need it.

Reply:
> yeah I guess I'd rather have an HCM option to delay route selection/tracing and let the filters do it and let this be a filter thing than invent another header modifier (but also again not blocking the plan, just suggesting alternates)

As I know, the route  (route-level upgrade map) is necesssary for HTTP filter chain creation. So, it's almost impossible to delay the route selection.

But I agree adding another header modifier is not best choice. I just couldn't figure out a better choice.

Reply:
I am thinking if is possible to add a special flag to the current http filter to tell it's a filter should run at very early.

And then we can extend the downstream `FilterManager` to add the methods to handle these early filters. We can add some limits to these filters, such as never stop, only for decoder, only for headers, no route dependency, etc.

And finally, we can implement a common filter to do the header mutations. This filter could run as early filter, normal downstream filter and even upstream filter. And the substitution formatter could be used in the filter. This filter even could be used to replace all `request_headers_to_add`/`response_headers_to_add`.

By this way, users could get a unified header modifier. 

But seems it's too complex for simple header mutation. (And I am not very sure if is possible to implement it as my above envision. Some pre-researchings are necessary.)

Reply:
Delaying route selection is almost impossible, especially considering that we may need enable different filters based on the route (very common requirement in the multi-tenant gateway) in the future.

Refactoring filter manager is too complex. It won't have better performance and users need understand the new concept of `early filter`.

May simple new extension point is an acceptable option. orz

But as @mattklein123 said, we can make the default extension more general (may be substitution formatter could be used here.)

Reply:
I'm OK with keeping the new extension point, but can we make it more general and potentially just allow the standard array of header options? Thank you.

/wait